vulnfeeds/vulns/vulns.go

```diff
@@ -259,6 +268,47 @@ func FromCVE(id string, cve cves.CVEItem, pkg, ecosystem, purl, versionType stri
 			URL:  reference.URL,
 		})
 	}
+	return &v, notes
+}
+
+func (v *Vulnerability) AddPkgInfo(pkgInfo []PackageInfo) {
+	for _, info := range pkgInfo {
+		affected := Affected{}
+		affected.Package.Name = info.PkgName
+		affected.Package.Ecosystem = info.Ecosystem
+		affected.Package.Purl = info.PURL
+		if info.FixedVersion != "" {
+			versionRange := AffectedRange{
+				Type: "ECOSYSTEM",
+				Events: []Event{
+					{Introduced: "0"},
+					{Fixed: info.FixedVersion},
+				},
+			}
+			affected.Ranges = append(affected.Ranges, versionRange)
+		}
+
+		if info.FixedCommit != "" {
+			versionRange := AffectedRange{
+				Type: "GIT",
+				Repo: info.Repo,
+				Events: []Event{
+					{Introduced: "0"},
+					{Fixed: info.FixedCommit},
+				},
+			}
+			affected.Ranges = append(affected.Ranges, versionRange)
+		}
+		v.Affected = append(v.Affected, affected)
+	}
+}
+
+// FromCVEWithVersionExtraction create an OSV object from a given CVEItem and package name
+// Tries to extract version information from the given CVEItem
+func FromCVEWithVersionExtraction(id string, cve cves.CVEItem, pkgInfo []PackageInfo, validVersions []string) (*Vulnerability, []string) {
+	v, notes := FromCVE(id, cve, pkgInfo)
+	// TODO(rexpan): Check this doesn't change behaviour
+	affected := &v.Affected[0]
```

Comment:
Rather than do this can you just add a boolean to control whether if we do version extraction? this can be part of the private interface i.e. `doFromCVE()`, and we can still have the public `FromCVE...` methods which take only the necessary args.

Reply:
Not too sure if that will fix the issue though, the new changes allows adding multiple PackageInfo, which means we still need the &v.Affected[0] line. It also shortens the very long FromCVE function into more modular parts.

Reply:
I see -- in that case, best to refactor this functionality into a separate function that's clearer about what it's doing. e.g.

```go
func ExtractVersionInformation(affected Affected, cve cves.CVEItem, ..)
```

We can also refactor `AddPkgInfo` to take a single PkgInfo instead, and return the converted `Affected` pointer, and remove that argument from `FromCVE`. 

That way, callers can be very explicit about the cases where they care about version extraction, rather than encoding this weird limitation here.

Something like...

```go
v, notes := vulns.FromCVE(id, cve)
affected := v.AddPkgInfo(vulns.PackageInfo{
  PkgName:   pkg,
  Ecosystem: "PyPI",
  PURL:      purl,
})
notes = append(notes, ExtractVersionInformation(affected, cve, ..., validVersions)...)
```


WDYT?

Reply:
That looks good to me, I'll refactor it now.

Reply:
It now looks something like this:
```
v, notes := vulns.FromCVE(id, cve)
v.AddPkgInfo(pkgInfo)

versions, versionNotes := cves.ExtractVersionInfo(cve, validVersions)
notes = append(notes, versionNotes...)

v.Affected[0].AttachExtractedVersionInfo(versions)
if len(v.Affected[0].Ranges) == 0 {
    log.Printf("No affected versions detected.")
}
```

A possibly cleaner version would be to have `addPkgInfo` return the affected field it added, but that would require changing the affected array in `Vulnerability` to an array of `*Affected` instead of `Affected`

Reply:
Thanks! This is fine for now. We can iterate on this later if you have better ideas here :)